File: WinForms/VB/AIReportDesignerDemo/Helpers/ConnectionHelper.vb
Imports System
Imports System.IO

Namespace DevExpress.AI.ReportDesigner.Demo

    Public Module ConnectionHelper

        Const dataDirectoryName As String = "DataDirectory"

        Friend Const DataDirectoryTag As String = "|" & dataDirectoryName & "|"

        Public Function PatchConnection(ByVal connectionString As String, ByVal dataDirectory As String) As String
            Return PatchConnectionCore(connectionString, Function() dataDirectory)
        End Function

        Public Function PatchConnectionCore(ByVal connectionString As String, ByVal getDataDirectory As Func(Of String)) As String
            If Not String.IsNullOrEmpty(connectionString) AndAlso connectionString.Contains(DataDirectoryTag) Then Return connectionString.Replace(DataDirectoryTag, getDataDirectory())
            Return connectionString
        End Function

        Private Function ValidateDirectory(ByVal dataDirectory As String) As String
            Dim result As String = Path.GetFullPath(dataDirectory)
            Return result.TrimEnd("\"c)
        End Function

        Public Function PatchDataDirectory(ByVal connectionString As String) As String
            Return PatchConnectionCore(connectionString, New Func(Of String)(AddressOf GetDataDirectory))
        End Function

        Private Function GetDataDirectory() As String
            Dim dataDirectory As String = TryCast(AppDomain.CurrentDomain.GetData(dataDirectoryName), String)
            If String.IsNullOrEmpty(dataDirectory) Then dataDirectory = AppDomain.CurrentDomain.BaseDirectory
            Return ValidateDirectory(dataDirectory)
        End Function

        Public Sub SetDataDirectory(ByVal directory As String)
            AppDomain.CurrentDomain.SetData("DataDirectory", directory)
        End Sub
    End Module
End Namespace

File: Reporting/CS/DevExpress.DemoReports/ConnectionHelper.cs
using System;
using System.IO;

namespace XtraReportsDemos {
    public static class ConnectionHelper {
        const string dataDirectoryName = "DataDirectory";
        internal const string DataDirectoryTag = "|" + dataDirectoryName + "|";

        public static string PatchConnection(string connectionString, string dataDirectory) {
            return PatchConnectionCore(connectionString, () => dataDirectory);
        }
        public static string PatchConnectionCore(string connectionString, Func<string> getDataDirectory) {
            if(!string.IsNullOrEmpty(connectionString) && connectionString.Contains(DataDirectoryTag))
                return connectionString.Replace(DataDirectoryTag, getDataDirectory());
            return connectionString;
        }
        static string ValidateDirectory(string dataDirectory) {
            string result = Path.GetFullPath(dataDirectory);
            return result.TrimEnd('\\');
        }
        public static string PatchDataDirectory(string connectionString) {
            return PatchConnectionCore(connectionString, GetDataDirectory);
        }
        static string GetDataDirectory() {
            string dataDirectory = AppDomain.CurrentDomain.GetData(dataDirectoryName) as string;
            if(string.IsNullOrEmpty(dataDirectory))
                dataDirectory = AppDomain.CurrentDomain.BaseDirectory;
            return ValidateDirectory(dataDirectory);
        }
        public static void SetDataDirectory(string directory) {
            AppDomain.CurrentDomain.SetData("DataDirectory", directory);
        }
    }
}
